pkg/virtctl/usbredir/client.go

```diff
@@ -0,0 +1,188 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright the KubeVirt Authors.
+ *
+ */
+
+package usbredir
+
+import (
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"os/exec"
+	"os/signal"
+	"time"
+
+	kvcorev1 "kubevirt.io/client-go/generated/kubevirt/clientset/versioned/typed/core/v1"
+	"kubevirt.io/client-go/kubecli"
+	"kubevirt.io/client-go/log"
+)
+
+type ClientConnectFn func(device, address string) error
+
+type Client struct {
+	// To connect local USB device buffer to the remote VM using the websocket.
+	inputReader  *io.PipeReader
+	inputWriter  *io.PipeWriter
+	outputReader *io.PipeReader
+	outputWriter *io.PipeWriter
+
+	listener *net.TCPListener
+
+	// channels
+	done   chan struct{}
+	stream chan error
+	local  chan error
+	remote chan error
+
+	ClientConnect ClientConnectFn
+}
+
+func NewUSBRedirClient() *Client {
+	inReader, inWriter := io.Pipe()
+	outReader, outWriter := io.Pipe()
+	return &Client{
+		inputReader:   inReader,
+		inputWriter:   inWriter,
+		outputReader:  outReader,
+		outputWriter:  outWriter,
+		ClientConnect: clientConnect,
+	}
+}
+
+func (k *Client) WithRemoteVMIStream(usbredirStream kubecli.StreamInterface) *Client {
+	k.stream = make(chan error)
+
+	go func() {
+		defer k.outputWriter.Close()
+		k.stream <- usbredirStream.Stream(
+			kvcorev1.StreamOptions{
+				In:  k.inputReader,
+				Out: k.outputWriter,
+			},
+		)
+	}()
+
+	return k
+}
+
+func (k *Client) WithLocalTCPClient(address string) *Client {
+	lnAddr, err := net.ResolveTCPAddr("tcp", address)
+	if err != nil {
+		log.Log.Errorf("Can't resolve the address: %s", err.Error())
+		return nil
+	}
+
+	// The local tcp server is used to proxy between remote websocket and local USB
+	k.listener, err = net.ListenTCP("tcp", lnAddr)
+	if err != nil {
+		log.Log.Errorf("Can't listen on unix socket: %s", err.Error())
+		return nil
```

Comment:
When I wrote this, I was actually thinking of using a file input instead of the actual TCP but in the end I didn't.
I'll move this to the factory function.

Reply:
It would be great if we can simplify this, I did not spend time looking at this but something like two callbacks returning some interface for `io.Copy` could be enough.

Reply:
Ok, I got what you mean. The only flexibility required at the moment is replacing the `ClientConnect` function, that the tests will use to mock data.  I'll reduce client's external functions (mostly making it not exported and re-used in the Factory, this way, it stays fragmented enough that we can rewrite later if needed).

Just note that my main goal is to share the main functionalities between what we use when running `virtctl usbredir` and `usbredir_test.go`. In the process I did get some things fixed, so, we are net positive ;)

Reply:
Did as mentioned in previous comment.